packages/app/src/cli/services/function/ui/components/hooks/setup-extension-watcher-for-replay.ts

```diff
@@ -0,0 +1,192 @@
+  const functionRunFromSelectedRun = {
+    type: 'functionRun',
+    input: selectedRun.payload.input,
+    output: selectedRun.payload.output,
+    logs: selectedRun.payload.logs,
+    name: selectedRun.source,
+    size: 0,
+    memory_usage: 0,
+    instructions: selectedRun.payload.fuelConsumed,
+  } as FunctionRun
+
+  const [logs, setLogs] = useState<ReplayLog[]>([])
+  const [recentFunctionRuns, setRecentFunctionRuns] = useState<[FunctionRun, FunctionRun]>([
+    functionRunFromSelectedRun,
+    functionRunFromSelectedRun,
+  ])
+
+  const [error, setError] = useState<string | undefined>(undefined)
+
+  const {input, export: runExport} = selectedRun.payload
+
+  const {isRawModeSupported: canUseShortcuts} = useStdin()
+  const pollingInterval = useRef<NodeJS.Timeout>()
+
+  const [statusMessage, setStatusMessage] = useState(`Watching for changes to ${selectedRun.source}...`)
+
+  useEffect(() => {
+    const initialReplay = async () => {
+      setStatusMessage('Replaying log with local function...')
+      const functionRun = await runFunctionRunnerWithLogInput(extension, JSON.stringify(input), runExport)
+      setRecentFunctionRuns((recentFunctionRuns) => {
+        return [functionRun, recentFunctionRuns[0]]
+      })
+      setStatusMessage(`Watching for changes to ${selectedRun.source}...`)
+      setLogs((logs) => [...logs, functionRun])
+    }
+
+    const startWatchingFunction = async () => {
+      const customStdout = new Writable({
+        write(chunk, _enconding, next) {
+          setLogs((logs) => [...logs, {type: 'systemMessage', message: chunk.toString()}])
+          next()
+        },
+      })
+
+      await setupExtensionWatcher({
+        extension,
+        app,
+        stdout: customStdout,
+        stderr: customStdout,
+        onChange: async () => {
+          setError('')
+          setStatusMessage('Re-running with latest changes...')
+          const functionRun = await runFunctionRunnerWithLogInput(extension, JSON.stringify(input), runExport)
+
+          setRecentFunctionRuns((recentFunctionRuns) => {
+            return [functionRun, recentFunctionRuns[0]]
+          })
+
+          setStatusMessage(`Watching for changes to ${selectedRun.source}...`)
+          setLogs((logs) => [...logs, functionRun])
+        },
+        onReloadAndBuildError: async (error) => {
+          setError('Error while reloading and building extension')
+          if (error instanceof FatalError) {
+            renderFatalError(error)
+          } else {
+            outputWarn(`Failed to replay function: ${error.message}`)
+          }
+        },
+        signal: abortController.signal,
+      })
+    }
+
+    // eslint-disable-next-line promise/catch-or-return, @typescript-eslint/no-floating-promises
+    initialReplay().then(() => {
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      startWatchingFunction()
+    })
+  }, [input, runExport, app, extension])
```

Comment:
Since this `useEffect` depends on these, it'll get re-run if any of them change. Since we don't return a clean up function, I think that means we'll end up with multiple watchers. To avoid this case, I think we should return a clean up function.

Reply:
Added this and also tested the cleanup manually